### PR TITLE
Gibbs Conditionals (new PR)

### PR DESCRIPTION
**Pull request by phipsgabler**
**This is just the same as https://github.com/TuringLang/Turing.jl/pull/1172, but I did some bad things with Git there, so creating a new PR seemed easier**

---

Introduces a sampler `GibbsConditional` that can be used to manually specify a -- surprise! -- Gibbs conditional for a variable:

```
sample(model, Gibbs(GibbsConditional(:λ, cond_λ), GibbsConditional(:m, cond_m)), N)
```

See also [this discussion](https://github.com/phipsgabler/LittleGibbs.jl/issues/3).

where the `cond_v` are functions from a `NamedTuple` of the conditioned variables (`(λ,)` for `m` and vice versa in the above example) to a `Distribution` (or probably just a `Sampleable` is enough).  [Example code](https://gist.github.com/phipsgabler/dfac25deceedd30c6642b8b6333797a0).

Currently, only one variable per sampler object is handled (as opposed to some other samplers).  Of course extension is possible, but I believe it is better to leave it this way and let `Gibbs` handle the sweeps.

Since this seemed too simple, I'm quite sure there's work left.  At least:

- [x] There's a lot of stuff (esp. `assume`/`observe`.) simply copied from MH -- can this be simplified or at least abstracted?
- [x] Documentation!
- [ ] Tests, naturally
  - [x] Basic feasibility (gdemo, mixture of Gaussians)
  - [ ] Check whether indexed variables really work
  - [x] Mixing Gibbs conditionals with other samplers
- [ ] Doctests (discuss first)

Currently, the code is working on the tests when only the conditionals or a combination of conditioals and MH is used, but fails in combination with an HMC sampler due to some AD things I don't understand.

### CODE REVIEW

File: src/inference/gibbs.jl [hunk 1]
```diff
@@ -166,27 +166,45 @@ function AbstractMCMC.step!(
...
 end
 
+
+"""
+    gibbs_step!(rng, model, spl, N, transition; kwargs...)
+
+Perform a conditional sampling step of the local sampler `spl` within a `Gibbs` sampler.  If the
+sampler is based on an `InferenceAlgorithm`, this just falls back to `AbstractMCMC.step!`.
+"""
+function gibbs_step!(
```
Comment (cpfiffer) [inline src/inference/gibbs.jl:196]: I'm not clear on why we use `gibbs_step!` instead of just `step!`. What's the value add here?

Comment (phipsgabler) [reply src/inference/gibbs.jl]: I honestely don't remember :) Presumably because `GibbsConditional` is not an `AbstractSampler`.  But removing it does not make it work not, so I did that.

Comment (cpfiffer) [reply src/inference/gibbs.jl]: Perfect, thanks!

Comment (devmotion) [reply src/inference/gibbs.jl]: It came up in https://github.com/TuringLang/Turing.jl/pull/1172#issuecomment-602189523. I don't remember exactly, but I think the idea was that you might want to perform some initial setup or configurations in the steps of the Gibbs sampler that you wouldn't perform in regular sampling. Currently that's all checked in the `step!` implementation (e.g., for HMC and ESS), but it would be cleaner to not mix these branches in the general `step!` implementation. The default implementation of `gibbs_step!` would just call `step!` (so not previous setup) but algorithms such as HMC could implement `gibbs_step!`, which would perform the setup and then call `step!`. Of course, `GibbsConditional` could then still just implement `step!` but I'm not sure if it makes sense to use GibbsConditional for anything else than Gibbs steps, so it might be a bit weird to keep it that general.

Comment (devmotion) [reply src/inference/gibbs.jl]: Maybe it would actually make sense to keep `gibbs_step!`? :thinking:

Comment (cpfiffer) [reply src/inference/gibbs.jl]: I feel like we should adhere to the interface as it currently is until we have a really good reason to deviate.

Comment (devmotion) [reply src/inference/gibbs.jl]: Yes, maybe it's actually better to clean the Gibbs specific parts of `step!` in a separate PR.

File: src/inference/gibbs_conditional.jl [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    conditional::C
+
+    function GibbsConditional(sym::Symbol, conditional::C) where {C}
+        return new{sym, C}(conditional)
+    end
+end
+
+DynamicPPL.getspace(::GibbsConditional{S}) where {S} = (S,)
+alg_str(::GibbsConditional) = "GibbsConditional"
+isgibbscomponent(::GibbsConditional) = true
```
Comment (devmotion) [inline src/inference/gibbs_conditional.jl:60]: I guess this should maybe also be part of another PR (since `GibbsComponent` is not removed yet)?
```suggestion
```

File: src/inference/gibbs_conditional.jl [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+"""Takes care of removing the `tonamedtuple` indexing form."""
+extractparam(p::Tuple{Vector{<:Array{<:Real}}, Vector{String}}) = foldl(vcat, p[1])
+function extractparam(p::Tuple{Vector{<:Real}, Vector{String}})
+    values, strings = p
+    if length(values) == length(strings) == 1 && !occursin(r".\[.+\]$", strings[1])
+        # if m ~ MVNormal(1, 1), we could have have ([1], ["m[1]"])!
+        return values[1]
+    else
+        return values
+    end
```
Comment (devmotion) [inline src/inference/gibbs_conditional.jl:133]: We should change the return type of `tonamedtuple` and not use strings (in a separate PR). In particular with the next MCMCChains release we would otherwise create strings in every step (when calling `tonamedtuple`) and then convert them back to symbols when creating the Chains object.

Comment (phipsgabler) [reply src/inference/gibbs_conditional.jl:133]: Yeah, given that the new `VarName`s are used already, this is a bit crazy.

Comment (devmotion) [reply src/inference/gibbs_conditional.jl:133]: I think we should leave that for a separate PR though, I assume it requires quite some changes that are otherwise completely unrelated to GibbsConditional.

File: test/inference/gibbs_conditional.jl [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            μ ~ arraydist(Normal.(fill(m, K), fill(λ, K)))
+            z ~ arraydist(Categorical.(fill(π, N)))
+            x ~ arraydist(Normal.(μ[z], σ))
+            return x
+        end
+
+        # see http://www.cs.columbia.edu/~blei/fogm/2015F/notes/mixtures-and-gibbs.pdf
+        function cond_z(c)
+            function mixtureweight(x)
+                p = π .* pdf.(Normal.(c.μ, σ), Ref(x))
```
Comment (mohamed82008) [inline test/inference/gibbs_conditional.jl:85]: Ref here is not needed?

File: src/inference/gibbs_conditional.jl [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    N::Integer,
+    transition;
+    kwargs...
+) where {S}
+    if spl.selector.rerun # Recompute joint in logp
+        model(spl.state.vi)
+    end
+
+    condvals = conditioned(tonamedtuple(spl.state.vi), Val{S}())
+    conddist = spl.alg.conditional(condvals)
```
Comment (mohamed82008) [inline src/inference/gibbs_conditional.jl:85]: This seems like a cool way to define a custom transition in general if we use all of the named tuple, not just the `condvals`. This would allow using the current value of the variable we are trying to sample to decide the next value, e.g. taking a gradient step. Without the MH criteria, this would not be exact MCMC but it can be useful to alternate between optimization and sampling for example. I would just pass the entire named tuple here.

**Comment by bdeonovic** [comment]
Whats the ETA of this ending up in master branch?

**Comment by phipsgabler** [comment]
@bdeonovic It's pretty much finished, and I'm bringing it up to current master right now.  There were always some strange things going on with AD, which is why it hasn't been merged yet (or rather, I stopped updating, since it basically was finished, but I'm not able to debug AD issues).

When I have it merged, can try it out on the current master, and I'd be happy to hear about any feedback.

File: src/inference/gibbs_conditional.jl (hunk 3 above)
Comment (phipsgabler) [reply src/inference/gibbs_conditional.jl:133]: Probably this will be done as part of what comes out from the discussion in https://github.com/TuringLang/DynamicPPL.jl/issues/119 over in DPPL.  Ideally, IMO, there would be `variable_values(vi) -> (x = ([1,2],), y = (3,))` and `variable_names(vi) ~> (x = (@varname(x[1]), @varname(x[2])), y = (@varname(y),))`, or something like this.

File: test/inference/gibbs_conditional.jl (hunk 4 above)
Comment (trappmartin) [reply test/inference/gibbs_conditional.jl:85]: I'd say it doesn't hurt to keep it.

File: src/inference/gibbs_conditional.jl (hunk 5 above)
Comment (phipsgabler) [reply src/inference/gibbs_conditional.jl:93]: Right, that's an interesting variant.  I did as you suggested.

**Comment by phipsgabler** [comment]
@bdeonovic OK, should be good now.  Combining a conditional with HMC still fails (not so dramatically anymore, but the results are wrong).

File: test/inference/gibbs_conditional.jl [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        (α_0, θ_0) = (2.0, inv(3.0)),
+        λ_true = rand(Gamma(α_0, θ_0)),
+        σ_true = √(1 / λ_true),
+        m_true = rand(Normal(0, σ_true)),
+        x = rand(Normal(m_true, σ_true), N)
+        
+        @model inverse_gdemo(x) = begin
+            λ ~ Gamma(α_0, θ_0)
+            m ~ Normal(0, sqrt(1 / λ))
+            x .~ Normal(m, sqrt(1 / λ))
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:18]: You have to use, e.g.,
```suggestion
            @. x ~ $(Normal(m, sqrt(1 / λ)))
```
or
```suggestion
            (.~)(x, Normal(m, sqrt(1 / λ)))
```
since `.~` can't be parsed by Julia 1.0 (see, e.g., https://github.com/TuringLang/DynamicPPL.jl/blob/06e7fe0cd08a22e728306a9444cf45d0e3ede690/test/compiler.jl#L440-L447).

**Comment by phipsgabler** [comment]
Back from zombie state :zombie: 

There's still the same error: the sampling of mu, in the GMM test case, fails with HMC using forwarddiff and reversediff.  No idea why.

Additionally, there needs to be a solution to make the test cases stable under permutation of cluster labels.  I'm not creative enough for that today... How is this usually done?

**Comment by trappmartin** [comment]
Does the same happen when you use the gdemo model and sample one parameter with HMC and the other through Gibbs sampling?

**Comment by phipsgabler** [comment]
@trappmartin Sometimes, but not consistently. And it also produces a huge pile of warnings:

```
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC ~/.julia/packages/AdvancedHMC/P9wqk/src/hamiltonian.jl:47
```

**Comment by trappmartin** [comment]
If the tests run through, this can be merged.

**Comment by trappmartin** [comment]
Looks like the test run through smoothly except for the multi threaded test, which terminates early due to time limit hit.

cc: @phipsgabler

**Comment by phipsgabler** [comment]
:+1: Seems like this can be merged.

@yebai apparently the logp-recalculation was not necessary, but my test cases were in fact broken.

Shall we do one more test with fancier indexing, or are [the current ones](https://github.com/phipsgabler/Turing.jl/blob/phg/new_gibbs_conditionals/test/inference/gibbs_conditional.jl) enough?

**Comment by devmotion** [comment]
Can you fix the Travis timeout?

The question is also if we want to merge this PR or https://github.com/TuringLang/Turing.jl/pull/1428 (updates Turing to AbstractMCMC 2 which has a different interface and hence requires different implementations of GibbsConditional) first. I have a deadline tomorrow but will try to finish the other PR (and the corresponding one to DynamicPPL) on Wednesday this week.

File: test/inference/gibbs.jl [hunk 7]
```diff
@@ -114,7 +114,7 @@ include(dir*"/test/test_utils/AllUtils.jl")
         alg = Gibbs(MH(:s), HMC(0.2, 4, :m))
         sample(model, alg, 100; callback = callback)
     end
-
+
```
Comment (devmotion) [inline test/inference/gibbs.jl:117]:
```suggestion

```

File: test/inference/gibbs.jl [hunk 8]
```diff
@@ -138,5 +138,5 @@ include(dir*"/test/test_utils/AllUtils.jl")
         model = imm(randn(100), 1.0);
         sample(model, Gibbs(MH(10, :z), HMC(0.01, 4, :m)), 100);
         sample(model, Gibbs(PG(10, :z), HMC(0.01, 4, :m)), 100);
-    end
+    end
```
Comment (devmotion) [inline test/inference/gibbs.jl:141]:
```suggestion
    end
```

**Comment by cpfiffer** [comment]
> The question is also if we want to merge this PR or #1428 (updates Turing to AbstractMCMC 2 which has a different interface and hence requires different implementations of GibbsConditional) first. I have a deadline tomorrow but will try to finish the other PR (and the corresponding one to DynamicPPL) on Wednesday this week.

I would say this one first since it's "smaller" in that it touches less of the code base.

File: test/inference/gibbs_conditional.jl [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+include(dir*"/test/test_utils/AllUtils.jl")
+
+
+@turing_testset "gibbs conditionals" begin
+    let N = 1000,
+        (α_0, θ_0) = (2.0, inv(3.0)),
+        λ_true = rand(Gamma(α_0, θ_0)),
+        σ_true = sqrt(1 / λ_true),
+        m_true = rand(Normal(0, σ_true)),
+        x = rand(Normal(m_true, σ_true), N)
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:14]: I'm curious, why a `let` block and not just regular assignments? Is it due to the other example below which uses variables of the same name? In this case two separate test sets might be an alternative.

File: test/inference/gibbs_conditional.jl [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        
+        @model function inverse_gdemo(x)
+            λ ~ Gamma(α_0, θ_0)
+            m ~ Normal(0, sqrt(1 / λ))
+            @. x ~ $(Normal(m, sqrt(1 / λ)))
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:19]: A bit simpler would be
```suggestion
            σ2 ~ InverseGamma(α_0, θ_0)
            σ = sqrt(σ2)
            m ~ Normal(0, σ)
            @. x ~ $(Normal(m, σ))
```

File: test/inference/gibbs_conditional.jl [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            x_bar = mean(x) # sample mean
+            s2 = var(x; mean=x_bar, corrected=false) # sample variance
+            return n, x_bar, s2
+        end
+
+        function cond_m(c)
+            n, x_bar, s2 = gdemo_statistics(x)
+            m_n = N * x_bar / (n + 1)
+            λ_n = c.λ * (n + 1)
+            σ_n = sqrt(1 / λ_n)
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:34]: Wouldn't it be easier to precompute `m_n` and a scaled variance outside of the function instead of closing over `x` and `N`? It is not clear if the compiler eliminates the redundant computations (it's also not possible for the square root).

File: test/inference/gibbs_conditional.jl [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+            return Normal(m_n, σ_n)
+        end
+
+        function cond_λ(c)
+            n, x_bar, s2 = gdemo_statistics(x)
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:39]: Same here, it seems one could precompute all required terms?

File: test/inference/gibbs_conditional.jl [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+            chain = sample(inverse_gdemo(x), alg, 10_000)
+            check_numerical(chain, [:m, :λ], [m_true, λ_true], atol=0.2)
+        end
+    end
+
+    Random.seed!(100)
+    let π = [0.5, 0.5],
+        K = length(π),
+        m = 0.5,
+        λ = 2.0,
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:63]: Again I'm curious why you use a `let` block here. It seems also a bit inconsistent/confusing to use `λ` for the precision above and for the standard deviation here.

File: test/inference/gibbs_conditional.jl [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        σ = 0.1,
+        N = 20,
+        μ_true = [rand(Normal(m, λ)) for _ in π],
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:66]:
```suggestion
        μ_true = rand(Normal(m, λ), length(π)),
```

File: test/inference/gibbs_conditional.jl [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        z_true = [rand(Categorical(π)) for _ in 1:N],
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:67]:
```suggestion
        z_true = rand(Categorical(π), N),
```

File: test/inference/gibbs_conditional.jl [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+        x = [rand(Normal(μ_true[z_n], σ)) for z_n in z_true]
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:68]:
```suggestion
        x = rand(MvNormal(μ_true[z_true], σ))
```

File: test/inference/gibbs_conditional.jl [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+        
+        @model function mixture(x)
+            μ ~ arraydist(Normal.(fill(m, K), fill(λ, K)))
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:71]:
```suggestion
            μ ~ MvNormal(Fill(m, K), λ)
```

File: test/inference/gibbs_conditional.jl [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+            z ~ arraydist(Categorical.(fill(π, N)))
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:72]:
```suggestion
            z ~ filldist(Categorical(π), N)
```

File: test/inference/gibbs_conditional.jl [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+            x ~ arraydist(Normal.(μ[z], σ))
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:73]:
```suggestion
            x ~ MvNormal(μ[z], σ)
```

File: test/inference/gibbs_conditional.jl [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+        function cond_μ(c)
+            z = c.z
+            n = [count(z .== k) for k = 1:K]
+
+            # If there were no observations assigned to center `k`, `n[k] == 0`, and
+            # we use the prior instead.
+            λ_hat = [(n[k] != 0) ? inv(n[k] / σ^2 + 1/λ^2) : λ for k = 1:K]
+            μ_hat = [(n[k] != 0) ? (sum(x[z .== k]) / σ^2) * λ_hat[k] : m for k = 1:K]
+
+            return arraydist(Normal.(μ_hat, λ_hat))
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:95]:
```suggestion
            return MvNormal(μ_hat, λ_hat)
```

File: test/inference/gibbs_conditional.jl [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+        # Both variables sampled using the Gibbs conditional
+        Random.seed!(100)
+        for alg in (Gibbs(GibbsConditional(:z, cond_z), GibbsConditional(:μ, cond_μ)),
+                    Gibbs(GibbsConditional(:z, cond_z), MH(:μ)),
+                    Gibbs(GibbsConditional(:z, cond_z), HMC(0.01, 7, :μ)), )
+
+            chain = sample(mixture(x), alg, 10000)
+            
+            μ_hat = estimate(chain, :μ)
+            lμ_hat, uμ_hat = minimum(μ_hat), maximum(μ_hat)
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:113]:
```suggestion
            lμ_hat, uμ_hat = extrema(μ_hat)
```

File: test/inference/gibbs_conditional.jl [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+            lμ_true, uμ_true = minimum(μ_true), maximum(μ_true)
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:114]: This could be precomputed outside of the loop it seems?
```suggestion
            lμ_true, uμ_true = extrema(μ_true)
```

File: src/inference/gibbs_conditional.jl [hunk 23]
```diff
@@ -0,0 +1,135 @@
+"""
+    GibbsConditional(sym, conditional)
+
+A "pseudo-sampler" to manually provide analytical Gibbs conditionals to `Gibbs`.
+`GibbsConditional(:x, cond)` will sample the variable `x` according to the conditional `cond`, which
+must therefore be a function from a `NamedTuple` of the conditioned variables to a `Distribution`.
+
+
+The `NamedTuple` that is passed in contains all random variables from the model in an unspecified
+order, taken from the [`VarInfo`](@ref) object over which the model is run.  Scalars and vectors are
```
Comment (devmotion) [inline src/inference/gibbs_conditional.jl:10]:
```suggestion
order, taken from the [`VarInfo`](@ref) object over which the model is run. Scalars and vectors are
```

File: src/inference/gibbs_conditional.jl [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+stored as their respective types, but all as floats.  The tuple also contains the value of the
```
Comment (devmotion) [inline src/inference/gibbs_conditional.jl:11]: Does the implementation require the limitation to floats?
```suggestion
stored as their respective types, but all as floats. The tuple also contains the value of the
```

File: src/inference/gibbs_conditional.jl [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+t = (m = ([0.234, -1.23], ["m[1]", "m[2]"]), λ = ([1.233], ["λ"])
+```
+
+and this function implements the cleanup of indexing. `conditioned(t)` will therefore return
+
+```julia
+(λ = 1.233, m = [0.234, -1.23])
+```
+"""
+@generated function conditioned(θ::NamedTuple{names}) where {names}
```
Comment (devmotion) [inline src/inference/gibbs_conditional.jl:119]: It seems this function (or maybe even rather a function that returns the same thing for a given VarInfo object directly without the intermediate `tonamedtuple`) could/should be added to DynamicPPL?

File: src/inference/gibbs_conditional.jl [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+
+# Examples
+
+```julia
+α_0 = 2.0
+θ_0 = inv(3.0)
+
+x = [1.5, 2.0]
+
+function gdemo_statistics(x)
```
Comment (devmotion) [inline src/inference/gibbs_conditional.jl:24]: I guess it would be reasonable to update the example as well?

File: test/inference/gibbs_conditional.jl [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+        σ_true = sqrt(1 / λ_true)
+        m_true = rand(Normal(0, σ_true))
+        x = rand(Normal(m_true, σ_true), N)
+
+        # The conditionals and posterior can be formulated in terms of the following statistics:
+        x_bar = mean(x) # sample mean
+        s2 = var(x; mean=x_bar, corrected=false) # sample variance
+        m_n = N * x_bar / (N + 1)
+        
+        @model function inverse_gdemo(x)
```
Comment (devmotion) [inline test/inference/gibbs_conditional.jl:24]: Is there a reason for not using the `gdemo` model directly (https://github.com/TuringLang/Turing.jl/blob/99ba1590dfbb121a38db55c8d9b5da8f2d4e22d4/test/test_utils/models.jl#L10-L16 - the main difference seems to be the number of observations?) or using an InverseGamma prior for the variance?

Comment (phipsgabler) [reply test/inference/gibbs_conditional.jl:24]: Yes: because I found it easier to express (or primarily, derive) the conditional in the precision parametrization.  But if someone cares to write conditionals for `gdemo`, I don't mind.

File: src/inference/gibbs_conditional.jl (hunk 23 above)
Comment (phipsgabler) [reply src/inference/gibbs_conditional.jl:10]: Just to mention that: the convention of separating sencences with double spaces didn't come naturally to me (it's an American idea, I have heard) -- but it makes navigation in Emacs easier, so I kind of do it automatically now :D

File: src/inference/gibbs_conditional.jl (hunk 24 above)
Comment (phipsgabler) [reply src/inference/gibbs_conditional.jl:11]: I believed it's a limitation of `VarInfo`, at least the untyped one -- is that wrong?

File: src/inference/gibbs_conditional.jl (hunk 23 above)
Comment (cpfiffer) [reply src/inference/gibbs_conditional.jl]: Lol no, so many Americans hate this too. There was kind of a big culture war about this recently. My preference is for single space between sentences, and I'd rather not choose spacing rules on the margin of "what makes emacs people happy".

File: src/inference/gibbs_conditional.jl (hunk 24 above)
Comment (devmotion) [reply src/inference/gibbs_conditional.jl]: The untyped `VarInfo` saves everything in a `Vector{Real}`: https://github.com/TuringLang/DynamicPPL.jl/blob/9431df05af76460b0d5fea9572fcaaf946b4791d/src/varinfo.jl#L160

Comment (phipsgabler) [reply src/inference/gibbs_conditional.jl]: OK, it probably is type-preserving with TypeVarInfo, but maybe not with the untyped one.  I removed the wording, it should rarely be a problem I guess.

**Comment by phipsgabler** [comment]
Should be fine now; there's still some issue with variable subsumption, but @trappmartin agreed that I should merge already.  I'll make a separate issue.

File: test/inference/gibbs_conditional.jl (hunk 27 above)
Comment (devmotion) [reply test/inference/gibbs_conditional.jl:24]: Aren't they basically the same, just using and return the inverse?

**Comment by devmotion** [comment]
You forgot to bump the version number - can you fix this in a new PR? :slightly_smiling_face:
